src/components/Map.js

```diff
@@ -1,229 +1,196 @@
+import React from 'react'
 import {RENDERING_CLUSTERS, RENDERING_HEATMAP, RENDERING_MARKERS, RENDERING_POLYLINE} from "./Layer";
 import L from 'leaflet';
 import 'leaflet.heat';
 import 'leaflet.markercluster';
 import 'leaflet.markercluster/dist/MarkerCluster.css';
 import 'leaflet.markercluster/dist/MarkerCluster.Default.css';
 
+/* 
+ * Main map component based on leaflet map.
+ */
+export const Map = React.memo(({layers}) => {
+
+	const mapRef = React.useRef();
+	const mapElementRef = React.createRef();
+
+	const mapOverlaysRef = React.useRef({});
+	const layerControlRef = React.useRef(new L.control.layers([]));
+
+	React.useEffect(() => {
+		const mapElement = mapElementRef.current;
+
+		if (mapElement) {
+			mapRef.current = L.map('map', {
+				preferCanvas: true,
+				center: [0, 0],
+				zoom: 2,
+				layers: [
+					L.tileLayer('http://{s}.tile.osm.org/{z}/{x}/{y}.png', {
+						attribution: '&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors'
+					}),
+				]
+			})
+		}
+
+	}, [mapElementRef])
 
-export class Map extends Component {
-
-	componentDidMount() {
-		// init an empty map
-		this.map = L.map('map', {
-			preferCanvas: true,
-			center: [0, 0],
-			zoom: 2,
-			layers: [
-				L.tileLayer('http://{s}.tile.osm.org/{z}/{x}/{y}.png', {
-					attribution: '&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors'
-				}),
-			]
-		});
-
-		this.leafletMarkerLayers = {};
-		this.leafletPolylineLayers = {};
-		this.leafletHeatmapLayers = {};
-		this.leafletClusterLayers = {};
-		this.layerControl = null;
-	}
-
-
-	componentDidUpdate() {
-		let layers = this.props.layers;
-		let globalBounds = new L.LatLngBounds();
-		let ukeyMarkerArray = [];
-		let ukeyPolylineArray = [];
-		let ukeyHeatmapArray = [];
-		let ukeyClusterArray = [];
-		// Iterate through layers
-		layers.map((layer) => {
-			if (layer.ukey === undefined)
-				return null;
-			let bds = new L.LatLngBounds(layer.bounds);
-			if (bds.isValid())
-				globalBounds.extend(bds);
-			if (layer.rendering === RENDERING_MARKERS) {
-				ukeyMarkerArray.push(layer.ukey);
-				if (!this.leafletMarkerLayers[layer.ukey]) {
-					this.leafletMarkerLayers[layer.ukey] = L.layerGroup().addTo(this.map);
-				}
-				this.updateMarkerLayer(layer.data, layer.color, layer.ukey);
-			} else if (layer.rendering === RENDERING_POLYLINE) {
-				ukeyPolylineArray.push(layer.ukey);
-				if (this.leafletPolylineLayers[layer.ukey]) {
-					// todo find a way of updating the polyline layer instead of delete & recreate
-					this.map.removeLayer(this.leafletPolylineLayers[layer.ukey]);
+	React.useEffect(() => {
+		const map = mapRef.current;
+
+		if (map) {
+
+			layerControlRef.current.remove(map);
+
+			let globalBounds = new L.LatLngBounds();
+
+			
+			// On a new render pass, build new map overlays object,
+			// and replace the current map overlays object. This auto-removes
+			// any deleted layers from map
+			const newMapOverlays = {}
```

Comment:
So, if I understand correctly, this overwrite all existing layers, right?

Reply:
Kind of building `newMapOverlays` from `oldMapOverlays`. OldMapOverlays are persisted across rerenders using the **mapOverlaysRef** . The map overlay already attached to map is not overwritten. All data within it gets overwritten at this time, but could be an update in future

Actual steps while iterating over the layers props are:

1. try get an old overlay from oldMapOverlays. `overlay = oldMapOverlays[layer.name]`. Create a new overlayif we found none. Overlay is (still) attached to map
2. update overlay by doing a data comparison and update or by clearing all data and adding fresh ones.
3. set this overlay in the `newMapOverlays`
4. swap out `newMapOverlays` for `newMapOverlays` in the **mapOverlaysRef**.

 I'm yet to validate these flows, so this PR is still a theoretic. The flow is uniform for all overlay rendering types